### PR TITLE
Add append element

### LINKED ISSUE #1801: Add append to contextPad
__What should we do?__

Add a "append anything" entry to the palette which provides a menu with all possible elements.

What it should look like:

 __To do:__
- [ ] Add entry to context pad 
- [ ] Add append menu entries provider

__Why should we do it?__

To support create anything, see https://github.com/bpmn-io/bpmn-js/issues/1627

### PR DESCRIPTION
**Pull request by smbea**
Closes #1801 
Depends on #1807

**How it looks:**

Uploading Screen Recording 2023-01-12 at 13.46.56.mov…

**Note:**
I used the same approach as for icons as we do in align and distribute. It seems to be the best solution we have for now (maybe the svg updates could be automated in the future)

### CODE REVIEW

**Comment by nikku** [comment]
@smbea Any reason we're not using the `...` icon? As far as I am aware the icon you chose is a web modeler extension, specifically to append a new connector. I'd not want to incorporate it into the core.

If you want to I can dig up earlier discussions on "which icon to choose", where we settled on "...".

File: lib/features/rules/BpmnRules.js [hunk 1]
```diff
@@ -801,6 +806,17 @@ function canResize(shape, newBounds) {
   return false;
 }
 
+function canAppend(source) {
+
+  if (isAny(source, [
+    'bpmn:EndEvent',
+    'bpmn:Group',
+    'bpmn:TextAnnotation',
+    'bpmn:SequenceFlow' ])) {
```
Comment (nikku) [inline lib/features/rules/BpmnRules.js:815]: Should it include participant, lane, data elements?

**Comment by philippfromme** [comment]
I found two quirks:

1. You can append elements that can only be connected in the opposite direction
2. You can append elements that cannot be connected at all

brave_brn6O8ga1Q

Are we aware of this?

**Comment by nikku** [comment]
@smbea Please also double-check existing bugs on the prototype. We want to make sure those don't re-open in the Camunda Modeler.

File: lib/features/context-pad/ContextPadProvider.js [hunk 2]
```diff
@@ -26,6 +26,8 @@ import {
   hasPrimaryModifier
 } from 'diagram-js/lib/util/Mouse';
 
+import ICONS from './ContextPadIcons';
```
Comment (nikku) [inline lib/features/context-pad/ContextPadProvider.js:29]: Feels kinda odd to now spread "icons" across the whole code base.

What about collecting all icons in a shared place? They can be bundled once, as needed, and consumed from anywhere.

I could imagine to provide icons through `lib/icons` to start with.

File: lib/features/context-pad/ContextPadIcons.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+/**
+ * To change the icons, modify the SVGs in `./resources`, execute `npx svgo -f resources --datauri enc -o dist`,
+ * and then replace respective icons with the optimized data URIs in `./dist`.
+ */
+var icons = {
```
Comment (nikku) [inline lib/features/context-pad/ContextPadIcons.js:5]: Let's not expose icons as a default export, but through a dedicated export:

```
const append = '...';

export { append };
``` 

This way the file can be tree-shaken (only actual icons used are consumed).

CC @barmac.

**Review by nikku — CHANGES_REQUESTED** [review]

File: lib/features/modeling/ElementFactory.js [hunk 4]
```diff
@@ -127,6 +127,10 @@ ElementFactory.prototype.createBpmnElement = function(elementType, attrs) {
     di.isMarkerVisible = true;
   }
 
+  if (attrs.triggeredByEvent) {
+    businessObject.triggeredByEvent = attrs.triggeredByEvent;
```
Comment (nikku) [inline lib/features/modeling/ElementFactory.js:131]: We should also delete `attrs` in this case, cf. handling of other special attributes.

Comment (nikku) [reply lib/features/modeling/ElementFactory.js:131]: Also, consider adding a test case for this.

**Review by nikku — APPROVED** [review]

**Review by nikku — CHANGES_REQUESTED** [review]
Great progress so far :tada:.

Some UX related findings:

* Append should only show prominent items initially (cf. [demo](https://bpmn-io.github.io/bpmn-js-connectors-extension/?aa=1)
  
  capture em6aBr_optimized

* Append should not offer blank `Boundary Event` => it is the same as an intermediate catch event
  
* Append should support _drag out to place anywhere_, (cf. [demo](https://bpmn-io.github.io/bpmn-js-connectors-extension/?aa=1):

  capture VWj1WG_optimized
  
* Append should allow to add boundary events (cf. [demo](https://bpmn-io.github.io/bpmn-js-connectors-extension/?aa=1), but other than in the demo **always attach manually**:
  
  capture KTMdrD_optimized

**Comment by nikku** [comment]
> In the spirit of integrating things incrementally, this can be implemented on top, no?

Yes, of course. Let's just ensure we don't at least evaluate it at some point.

I've added it to the [EPIC](https://github.com/bpmn-io/bpmn-js/issues/1627) just to be sure.

File: lib/features/popup-menu/AppendMenuProvider.js [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+};
+
+/**
+ * Get all entries from createOptions for the given element.
+ *
+ * @param {djs.model.Base} element
+ *
+ * @return {Array<Object>} a list of menu entry items
+ */
+AppendMenuProvider.prototype.getEntries = function(element) {
```
Comment (nikku) [inline lib/features/popup-menu/AppendMenuProvider.js:49]: Let's use the new version `PopupMenuProvider#getPopupMenuEntries`.

```suggestion
AppendMenuProvider.prototype.getPopupMenuEntries = function(element) {
```

**Review by nikku — APPROVED** [review]
Thanks for this contribution! Very nice:

* Integration with rules
* Improved handling of icons
* Solid test coverage
* Really intuitive modeling

----

To highlight one thing I love how natural this feels to me (auto place vs. manual placement), i.e. considering the "non-interrupting boundary" placement here:

capture eoOWhe_optimized

----

If I had one wish then it would be to be less strict regarding when we show the menu. i.e. for link throw events:

screenshot DZKHtk

If we combine "showing it in more places" with https://github.com/bpmn-io/bpmn-js/issues/1621 you could argue that creating a (matching) link catch event (or creating a matching catching event for whatever throw event is selected) could be the primary action provided to a user. :)

Great work overall.

**Pull request merged**